### ISSUE TITLE
An error is reported while running the keras model：The input to the layer needs to be a vector(or a mini-batch of vectors);

### ISSUE THREAD

**Issue by 540667387** [intel-analytics/BigDL#3127]
21/05/21 16:39:40 INFO spark.ContextCleaner: Cleaned accumulator 207
21/05/21 16:39:40 INFO codegen.CodeGenerator: Code generated in 14.815498 ms
21/05/21 16:39:41 INFO optim.DistriOptimizer$: caching training rdd ...
21/05/21 16:39:43 INFO optim.DistriOptimizer$: Cache thread models...
21/05/21 16:39:44 INFO optim.DistriOptimizer$: Cache thread models... done
21/05/21 16:39:44 INFO optim.DistriOptimizer$: Count dataset
21/05/21 16:39:45 INFO optim.DistriOptimizer$: Count dataset complete. Time elapsed: 0.183908219s
21/05/21 16:39:45 INFO optim.DistriOptimizer$: config  {
        computeThresholdbatchSize: 100
        maxDropPercentage: 0.0
        warmupIterationNum: 200
        isLayerwiseScaled: false
        dropPercentage: 0.0
 }
21/05/21 16:39:45 INFO optim.DistriOptimizer$: Shuffle data
21/05/21 16:39:45 INFO optim.DistriOptimizer$: Shuffle data complete. Takes 0.02642685s
2021-05-21 16:39:45 ERROR TaskSetManager:70 - Task 6 in stage 60.0 failed 1 times; aborting job
21/05/21 16:39:45 ERROR optim.DistriOptimizer$: Error: org.apache.spark.SparkException: Job aborted due to stage failure: Task 6 in stage 60.0 failed 1 times, most recent failure: Lost task 6.0 in stage 60.0 (TID 5602, cdh-slave13.nielsenhypers.com, executor 4): Layer info: Sequential[a7fd4950]/Embedding[a1b53bd9]/LookupTable[2f2f6fa3](nIndex=3163,nOutput=256,paddingValue=0.0,normType=2.0)
java.lang.IllegalArgumentException: requirement failed: LookupTable:
 The input to the layer needs to be a vector(or a mini-batch of vectors);
 please use the Reshape module to convert multi-dimensional input into vectors
 if appropriate"
    , input dim [3]
        at scala.Predef$.require(Predef.scala:224)
        at com.intel.analytics.bigdl.nn.LookupTable.updateOutput(LookupTable.scala:174)
        at com.intel.analytics.bigdl.nn.LookupTable.updateOutput(LookupTable.scala:47)
        at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:282)
        at com.intel.analytics.bigdl.nn.Sequential.updateOutput(Sequential.scala:39)
        at com.intel.analytics.bigdl.nn.keras.KerasLayer.updateOutput(KerasLayer.scala:274)
        at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:282)
        at com.intel.analytics.bigdl.nn.Sequential.updateOutput(Sequential.scala:39)
        at com.intel.analytics.bigdl.nn.keras.KerasLayer.updateOutput(KerasLayer.scala:274)
        at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:282)
        at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply$mcI$sp(DistriOptimizer.scala:269)
        at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply(DistriOptimizer.scala:260)
        at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply(DistriOptimizer.scala:260)
        at com.intel.analytics.bigdl.utils.ThreadPool$$anonfun$1$$anon$5.call(ThreadPool.scala:160)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

        at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:288)
        at com.intel.analytics.bigdl.nn.Sequential.updateOutput(Sequential.scala:39)
        at com.intel.analytics.bigdl.nn.keras.KerasLayer.updateOutput(KerasLayer.scala:274)
        at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:282)
        at com.intel.analytics.bigdl.nn.Sequential.updateOutput(Sequential.scala:39)
        at com.intel.analytics.bigdl.nn.keras.KerasLayer.updateOutput(KerasLayer.scala:274)
        at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:282)
        at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply$mcI$sp(DistriOptimizer.scala:269)
        at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply(DistriOptimizer.scala:260)
        at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply(DistriOptimizer.scala:260)
        at com.intel.analytics.bigdl.utils.ThreadPool$$anonfun$1$$anon$5.call(ThreadPool.scala:160)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

I guess my input shape is wrong, but I'm not sure？

**Comment by dding3** [comment]
Yes, from the exception "java.lang.IllegalArgumentException: requirement failed: LookupTable:
The input to the layer needs to be a vector(or a mini-batch of vectors);
please use the Reshape module to convert multi-dimensional input into vectors
if appropriate"
, input dim [3]",
seems the input to the LookupTable is not correct. 

You may want to refer https://bigdl-project.github.io/master/#APIGuide/Layers/Embedding-Layers/#lookuptable for the detailed usage.

**Comment by jason-dai** [comment]
> 21/05/21 16:39:40 INFO spark.ContextCleaner: Cleaned accumulator 207
> 21/05/21 16:39:40 INFO codegen.CodeGenerator: Code generated in 14.815498 ms
> 21/05/21 16:39:41 INFO optim.DistriOptimizer$: caching training rdd ...
> 21/05/21 16:39:43 INFO optim.DistriOptimizer$: Cache thread models...
> 21/05/21 16:39:44 INFO optim.DistriOptimizer$: Cache thread models... done
> 21/05/21 16:39:44 INFO optim.DistriOptimizer$: Count dataset
> 21/05/21 16:39:45 INFO optim.DistriOptimizer$: Count dataset complete. Time elapsed: 0.183908219s
> 21/05/21 16:39:45 INFO optim.DistriOptimizer$: config {
> computeThresholdbatchSize: 100
> maxDropPercentage: 0.0
> warmupIterationNum: 200
> isLayerwiseScaled: false
> dropPercentage: 0.0
> }
> 21/05/21 16:39:45 INFO optim.DistriOptimizer$: Shuffle data
> 21/05/21 16:39:45 INFO optim.DistriOptimizer$: Shuffle data complete. Takes 0.02642685s
> 2021-05-21 16:39:45 ERROR TaskSetManager:70 - Task 6 in stage 60.0 failed 1 times; aborting job
> 21/05/21 16:39:45 ERROR optim.DistriOptimizer$: Error: org.apache.spark.SparkException: Job aborted due to stage failure: Task 6 in stage 60.0 failed 1 times, most recent failure: Lost task 6.0 in stage 60.0 (TID 5602, cdh-slave13.nielsenhypers.com, executor 4): Layer info: Sequential[a7fd4950]/Embedding[a1b53bd9]/LookupTable[2f2f6fa3](nIndex=3163,nOutput=256,paddingValue=0.0,normType=2.0)
> java.lang.IllegalArgumentException: requirement failed: LookupTable:
> The input to the layer needs to be a vector(or a mini-batch of vectors);
> please use the Reshape module to convert multi-dimensional input into vectors
> if appropriate"
> , input dim [3]
> at scala.Predef$.require(Predef.scala:224)
> at com.intel.analytics.bigdl.nn.LookupTable.updateOutput(LookupTable.scala:174)
> at com.intel.analytics.bigdl.nn.LookupTable.updateOutput(LookupTable.scala:47)
> at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:282)
> at com.intel.analytics.bigdl.nn.Sequential.updateOutput(Sequential.scala:39)
> at com.intel.analytics.bigdl.nn.keras.KerasLayer.updateOutput(KerasLayer.scala:274)
> at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:282)
> at com.intel.analytics.bigdl.nn.Sequential.updateOutput(Sequential.scala:39)
> at com.intel.analytics.bigdl.nn.keras.KerasLayer.updateOutput(KerasLayer.scala:274)
> at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:282)
> at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply$mcI$sp(DistriOptimizer.scala:269)
> at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply(DistriOptimizer.scala:260)
> at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply(DistriOptimizer.scala:260)
> at com.intel.analytics.bigdl.utils.ThreadPool$$anonfun$1$$anon$5.call(ThreadPool.scala:160)
> at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> at java.lang.Thread.run(Thread.java:748)
> 
> ```
>     at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:288)
>     at com.intel.analytics.bigdl.nn.Sequential.updateOutput(Sequential.scala:39)
>     at com.intel.analytics.bigdl.nn.keras.KerasLayer.updateOutput(KerasLayer.scala:274)
>     at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:282)
>     at com.intel.analytics.bigdl.nn.Sequential.updateOutput(Sequential.scala:39)
>     at com.intel.analytics.bigdl.nn.keras.KerasLayer.updateOutput(KerasLayer.scala:274)
>     at com.intel.analytics.bigdl.nn.abstractnn.AbstractModule.forward(AbstractModule.scala:282)
>     at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply$mcI$sp(DistriOptimizer.scala:269)
>     at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply(DistriOptimizer.scala:260)
>     at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$4$$anonfun$5$$anonfun$apply$2.apply(DistriOptimizer.scala:260)
>     at com.intel.analytics.bigdl.utils.ThreadPool$$anonfun$1$$anon$5.call(ThreadPool.scala:160)
>     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
>     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
>     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
>     at java.lang.Thread.run(Thread.java:748)
> ```
> 
> I guess my input shape is wrong, but I'm not sure？

I would recommend using the high level Keras-like API for BigDL (https://analytics-zoo.readthedocs.io/en/latest/doc/UseCase/keras-api.html), which automatically infers the shape for the layers.

**Comment by helenlly** [comment]
@540667387 any more questions on the issue?
